File: src/freight_fate/data/world_services.py
"""Source-backed city service and local route helpers for world data."""

from __future__ import annotations

import zlib

from .world_constants import (
    CITY_SERVICE_ORDER,
    CITY_SERVICE_SOURCE_NOTES,
    FACILITY_APPROACH_MILES,
    FACILITY_APPROACH_ROADS,
)
from .world_models import (
    CityService,
    FacilityApproach,
    FacilityEndpoint,
    Leg,
    LocalApproach,
    LocalGeometry,
    Route,
)

CITY_SERVICE_APPROACH_MILES = {
    "freight_market": 3.0,
    "garage": 1.5,
    "truck_dealer": 2.5,
}

CITY_SERVICE_APPROACH_ROADS = {
    "freight_market": "local freight office access road",
    "garage": "terminal service lane",
    "truck_dealer": "dealer access road",
}


class WorldServiceMixin:
    def city_services(self, city: str) -> tuple[CityService, ...]:
        """Service POIs available for local city driving.

        Source-backed entries from ``city_services.json`` are preferred per
        service key. Missing keys stay available as representative fallback
        services so the existing offline menu contract remains complete.
        ``CityService.city`` carries the canonical city key so it round-trips
        through the other service lookups; spoken text uses ``name``.
        """
        city_key = self.resolve_city_key(city)
        if city_key not in self.cities:
            raise KeyError(f"Unknown city: {city}")
        source_entries = self._city_service_data.get(city_key, {})
        services: list[CityService] = []
        for key in CITY_SERVICE_ORDER:
            raw = source_entries.get(key)
            if raw is None:
                services.append(self._fallback_city_service(city_key, key))
                continue
            city_obj = self.cities[city_key]
            services.append(
                CityService(
                    key=key,
                    name=str(raw["name"]).strip(),
                    city=city_key,
                    state=city_obj.state,
                    kind=str(raw.get("kind", key)).strip() or key,
                    source_note=str(raw.get("source_note", "")).strip(),
                    lat=float(raw.get("lat", 0.0)),
                    lon=float(raw.get("lon", 0.0)),
                    approach_miles=round(float(raw["approach_miles"]), 1),
                    approach_road=str(raw["approach_road"]).strip(),
                    source_type=str(raw.get("source_type", "osm")).strip(),
                    source_ref=str(raw.get("source_ref", "")).strip(),
                    fallback=bool(raw.get("fallback", False)),
                    fallback_reason=str(raw.get("fallback_reason", "")).strip(),
                )
            )
        return tuple(services)

    def _fallback_city_service(self, city_key: str, key: str) -> CityService:
        city_obj = self.cities[city_key]
        terminal = self.home_terminal(city_key)
        names = {
            "freight_market": f"{city_obj.name} Freight Market Office",
            "garage": f"{terminal.name} Garage",
            "truck_dealer": f"{city_obj.name} Truck Dealer",
        }
        return CityService(
            key=key,
            name=names[key],
            city=city_key,
            state=city_obj.state,
            kind=key,
            source_note=CITY_SERVICE_SOURCE_NOTES[key],
            fallback_reason="No checked-in source-backed city service entry for this role.",
        )

    def city_service(self, city: str, key: str) -> CityService:
        for service in self.city_services(city):
            if service.key == key:
                return service
        raise KeyError(f"Unknown service in {city}: {key}")

    def local_approach(self, target_id: str) -> LocalApproach | None:
        return self._local_approaches.get(target_id)

    def local_geometry(self, target_id: str) -> LocalGeometry | None:
        return self._local_geometries.get(target_id)

    def city_service_approach(self, city: str, key: str) -> LocalApproach | None:
        return self.local_approach(f"city_service:{self.resolve_city_key(city)}:{key}")

    def city_service_geometry(self, city: str, key: str) -> LocalGeometry | None:
        return self.local_geometry(f"city_service:{self.resolve_city_key(city)}:{key}")

    def facility_approach(self, city: str, location_name: str) -> LocalApproach | None:
        location = self.facility_location(city, location_name)
        return self.local_approach(f"facility:{location.id}")

    def facility_endpoint(self, city: str, location_name: str) -> FacilityEndpoint | None:
        location = self.facility_location(city, location_name)
        return self._facility_endpoints.get(location.id)

    def facility_source_approach(self, city: str, location_name: str) -> FacilityApproach | None:
        location = self.facility_location(city, location_name)
        return self._facility_approaches.get(location.id)

    def facility_geometry(self, city: str, location_name: str) -> LocalGeometry | None:
        location = self.facility_location(city, location_name)
        return self.local_geometry(f"facility:{location.id}")

    def city_service_route(self, city: str, key: str) -> Route:
        """A short, drivable local route from the terminal to a city service."""
        city = self.resolve_city_key(city)
        service = self.city_service(city, key)
        geometry = self.city_service_geometry(city, key)
        if geometry is not None and geometry.turn_level and geometry.segments:
            legs = [
                Leg(
                    city,
                    city,
                    segment.miles,
                    segment.road,
                    "flat",
                    (),
                    local_cue=segment.cue,
                    local_speed_mph=segment.speed_mph,
                )
                for segment in geometry.segments
            ]
            return Route([city] * (len(legs) + 1), legs)
        approach = self.city_service_approach(city, key)
        if approach is not None:
            miles = approach.approach_miles
            road = approach.road
        elif service.approach_miles > 0:
            miles = service.approach_miles
            road = service.approach_road or CITY_SERVICE_APPROACH_ROADS.get(
                service.kind, "city service road"
            )
        else:
            base_miles = CITY_SERVICE_APPROACH_MILES.get(service.kind, 3.0)
            seed = zlib.crc32(f"{city}:service:{service.key}".encode())
            offset = (seed % 5) * 0.2
            miles = round(base_miles + offset, 1)
            road = CITY_SERVICE_APPROACH_ROADS.get(service.kind, "city service road")
        leg = Leg(city, city, miles, road, "flat", ())
        return Route([city, city], [leg])

    def facility_approach_route(self, city: str, location_name: str) -> Route:
        """A short, drivable local route from the company terminal to a facility."""
        city = self.resolve_city_key(city)
        location = self.facility_location(city, location_name)
        source_approach = self._facility_approaches.get(location.id)
        if source_approach is not None and source_approach.turn_level and source_approach.segments:
            legs = [
                Leg(
                    city,
                    city,
                    segment.miles,
                    segment.road,
                    "flat",
                    (),
                    local_cue=segment.cue,
                    local_speed_mph=segment.speed_mph,
                )
                for segment in source_approach.segments
            ]
            return Route([city] * (len(legs) + 1), legs)
        endpoint = self._facility_endpoints.get(location.id)
        approach = self.local_approach(f"facility:{location.id}")
        if endpoint is not None and endpoint.source_backed and not endpoint.fallback:
            miles = endpoint.approach_miles
            road = approach.road if approach is not None else endpoint.approach_road
        elif approach is not None:
            miles = approach.approach_miles
            road = approach.road
        else:
            base_miles = FACILITY_APPROACH_MILES.get(location.type, 4.0)
            seed = zlib.crc32(f"{city}:{location.name}:{location.type}".encode())
            offset = (seed % 7) * 0.25
            miles = round(base_miles + offset, 1)
            road = FACILITY_APPROACH_ROADS.get(location.type, "facility access road")
        leg = Leg(city, city, miles, road, "flat", ())
        return Route([city, city], [leg])

File: crates/ff-core/src/data/world_services.rs
//! Source-backed city service and local route helpers for world data (port of
//! `freight_fate/data/world_services.py` -- Python's `WorldServiceMixin`,
//! here a second `impl World` block).

use once_cell::sync::Lazy;
use regex::Regex;

use super::world::World;
use super::world_constants::{
    lookup, CITY_SERVICE_ORDER, CITY_SERVICE_SOURCE_NOTES, FACILITY_APPROACH_MILES,
    FACILITY_APPROACH_ROADS, FACILITY_APPROACH_TRUSTED_MAX_MI,
};
use super::world_models::{
    CityService, DataError, FacilityApproach, FacilityEndpoint, Leg, LocalApproach, LocalGeometry,
    Route,
};
use super::world_parsing::crc32;
use crate::pyfmt::round_py_n;

static ROAD_REF_LIST: Lazy<Regex> =
    Lazy::new(|| Regex::new(r"\(([^()]*;[^()]*)\)").expect("valid regex"));

/// Josh's ruling (2026-07-24): local deadheads run 1 to 9 miles. The bake
/// already floors facilities at 2.1; this caps the synthetic single-leg
/// top end until the placement audit re-geocodes the misplaced pins.
pub const SYNTHETIC_APPROACH_CAP_MI: f64 = 9.0;

/// Trim raw OSM ref lists out of player-facing street text.
///
/// Source-backed street names sometimes carry the full multi-ref
/// parenthetical straight from the map tags -- "North Michigan Street
/// (SR 933;BUS US 31)". Read aloud, the semicolon list is tag soup, so
/// keep the first ref only: "North Michigan Street (SR 933)".
pub fn spoken_road_text(text: &str) -> String {
    if text.is_empty() || !text.contains(';') {
        return text.to_string();
    }
    ROAD_REF_LIST
        .replace_all(text, |caps: &regex::Captures| {
            let first = caps[1].split(';').next().unwrap_or("").trim().to_string();
            format!("({first})")
        })
        .into_owned()
}

/// Maneuver direction baked in a local segment cue, or "".
fn local_cue_direction(cue: &str) -> &'static str {
    let lowered = cue.trim().to_lowercase();
    if lowered.starts_with("turn left") {
        "left"
    } else if lowered.starts_with("turn right") {
        "right"
    } else if lowered.starts_with("continue") {
        "ahead"
    } else {
        ""
    }
}

/// The same street chain driven outbound: leg order reversed, and each
/// junction's turn direction flipped (an inbound right turn is an outbound
/// left at the same corner). Near-straight boundaries stay "Continue onto";
/// directionless legacy cues stay directionless.
fn reversed_local_legs(city: &str, legs: &[std::sync::Arc<Leg>]) -> Vec<Leg> {
    let mut out = Vec::with_capacity(legs.len());
    for (i, src) in legs.iter().rev().enumerate() {
        let cue = if i == 0 {
            format!("Start on {}.", src.highway)
        } else {
            // Outbound, the junction onto this leg is the one the inbound
            // drive crossed *leaving* it: the cue baked on the leg after it.
            let inbound = local_cue_direction(&legs[legs.len() - i].local_cue);
            match inbound {
                "left" => format!("Turn right onto {}.", src.highway),
                "right" => format!("Turn left onto {}.", src.highway),
                "ahead" => format!("Continue onto {}.", src.highway),
                _ => format!("Turn onto {}.", src.highway),
            }
        };
        out.push(Leg::local(
            city,
            src.miles,
            &src.highway,
            &cue,
            src.local_speed_mph,
        ));
    }
    out
}

impl World {
    /// Service POIs available for local city driving.
    ///
    /// Source-backed entries from `city_services.json` are preferred per
    /// service key. Missing keys stay available as representative fallback
    /// services so the existing offline menu contract remains complete.
    /// `CityService.city` carries the canonical city key so it round-trips
    /// through the other service lookups; spoken text uses `name`.
    pub fn city_services(&self, city: &str) -> Result<Vec<CityService>, DataError> {
        let city_key = self.resolve_city_key(city);
        let Some(city_obj) = self.cities.get(&city_key) else {
            return Err(DataError::key(format!("Unknown city: {city}")));
        };
        let source_entries = self.city_service_data()?.get(&city_key);
        let mut services = Vec::with_capacity(CITY_SERVICE_ORDER.len());
        for key in CITY_SERVICE_ORDER {
            let Some(raw) = source_entries.and_then(|entries| entries.get(*key)) else {
                services.push(self.fallback_city_service(&city_key, key)?);
                continue;
            };
            let mut kind = raw.kind.as_deref().unwrap_or(key).trim().to_string();
            if kind.is_empty() {
                kind = key.to_string();
            }
            services.push(CityService {
                key: key.to_string(),
                name: raw.name.trim().to_string(),
                city: city_key.clone(),
                state: city_obj.state.clone(),
                kind,
                source_note: raw.source_note.trim().to_string(),
                lat: raw.lat,
                lon: raw.lon,
                approach_miles: round_py_n(raw.approach_miles, 1),
                approach_road: raw.approach_road.trim().to_string(),
                source_type: raw
                    .source_type
                    .as_deref()
                    .unwrap_or("osm")
                    .trim()
                    .to_string(),
                source_ref: raw.source_ref.trim().to_string(),
                fallback: raw.fallback,
                fallback_reason: raw.fallback_reason.trim().to_string(),
            });
        }
        Ok(services)
    }

    fn fallback_city_service(&self, city_key: &str, key: &str) -> Result<CityService, DataError> {
        let city_obj = &self.cities[city_key];
        let terminal = self.home_terminal(city_key)?;
        let name = match key {
            "freight_market" => format!("{} Freight Market Office", city_obj.name),
            "garage" => format!("{} Garage", terminal.name),
            "truck_dealer" => format!("{} Truck Dealer", city_obj.name),
            other => return Err(DataError::key(py_key(other))),
        };
        Ok(CityService {
            key: key.to_string(),
            name,
            city: city_key.to_string(),
            state: city_obj.state.clone(),
            kind: key.to_string(),
            source_note: lookup(CITY_SERVICE_SOURCE_NOTES, key)
                .unwrap_or_default()
                .to_string(),
            fallback_reason: "No checked-in source-backed city service entry for this role."
                .to_string(),
            ..CityService::default()
        })
    }

    pub fn city_service(&self, city: &str, key: &str) -> Result<CityService, DataError> {
        self.city_services(city)?
            .into_iter()
            .find(|service| service.key == key)
            .ok_or_else(|| DataError::key(format!("Unknown service in {city}: {key}")))
    }

    pub fn local_approach(&self, target_id: &str) -> Result<Option<&LocalApproach>, DataError> {
        Ok(self.local_approaches()?.get(target_id))
    }

    pub fn local_geometry(&self, target_id: &str) -> Result<Option<&LocalGeometry>, DataError> {
        Ok(self.local_geometries()?.get(target_id))
    }

    pub fn facility_approach(
        &self,
        city: &str,
        location_name: &str,
    ) -> Result<Option<&LocalApproach>, DataError> {
        let location = self.facility_location(city, location_name)?;
        self.local_approach(&format!("facility:{}", location.id))
    }

    pub fn facility_endpoint(
        &self,
        city: &str,
        location_name: &str,
    ) -> Result<Option<&FacilityEndpoint>, DataError> {
        let location = self.facility_location(city, location_name)?;
        Ok(self.facility_endpoints()?.get(&location.id))
    }

    pub fn facility_source_approach(
        &self,
        city: &str,
        location_name: &str,
    ) -> Result<Option<&FacilityApproach>, DataError> {
        let location = self.facility_location(city, location_name)?;
        Ok(self.facility_approaches()?.get(&location.id))
    }

    /// Local approach road a HIGHWAY run has to cover to reach this gate.
    ///
    /// The arrival zones size the destination approach from this rather than
    /// from a flat mileage: the facilities differ hugely, and a number that
    /// fits a dock two ramps off the interstate is a crawl for one sitting on
    /// the frontage road.
    ///
    /// `None` means "no usable geometry, size it synthetically", and it is
    /// the answer in two different cases. A facility with a genuine
    /// turn-level street chain has that chain driven as a route of its own
    /// once the highway run ends, so counting its mileage here as well would
    /// slow the freeway for road the truck has not reached. A facility whose
    /// record is a fallback, or whose endpoint estimate is longer than any
    /// real approach road (the misplaced pins the synthetic cap already
    /// guards), has nothing worth believing.
    pub fn facility_approach_miles(
        &self,
        city: &str,
        location_name: &str,
    ) -> Result<Option<f64>, DataError> {
        let Ok(location) = self.facility_location(city, location_name) else {
            return Ok(None);
        };
        let approach = self.facility_approaches()?.get(&location.id);
        if let Some(approach) = approach {
            if approach.turn_level && !approach.segments.is_empty() {
                return Ok(None); // its own street chain covers this road
            }
            if !approach.fallback && approach.total_miles > 0.0 {
                return Ok(Some(approach.total_miles));
            }
        }
        if let Some(endpoint) = self.facility_endpoints()?.get(&location.id) {
            if endpoint.source_backed && !endpoint.fallback {
                let miles = endpoint.approach_miles;
                if 0.0 < miles && miles <= FACILITY_APPROACH_TRUSTED_MAX_MI {
                    return Ok(Some(miles));
                }
            }
        }
        Ok(None)
    }

    pub fn facility_geometry(
        &self,
        city: &str,
        location_name: &str,
    ) -> Result<Option<&LocalGeometry>, DataError> {
        let location = self.facility_location(city, location_name)?;
        self.local_geometry(&format!("facility:{}", location.id))
    }

    /// The facility's street chain driven outbound -- gate toward the
    /// highway on-ramp -- or `None` when the facility has no genuine
    /// multi-segment turn-level chain (those keep the scripted departure).
    /// Mirrors the arrival-side chain gating in the driving layer.
    pub fn facility_departure_route(
        &self,
        city: &str,
        location_name: &str,
    ) -> Result<Option<Route>, DataError> {
        let route = self.facility_approach_route(city, location_name)?;
        if route.legs.len() < 2 {
            return Ok(None);
        }
        if !route.legs.iter().any(|leg| leg.local_speed_mph > 0.0) {
            return Ok(None);
        }
        let city = self.resolve_city_key(city);
        let legs = reversed_local_legs(&city, &route.legs);
        let cities = vec![city; legs.len() + 1];
        Ok(Some(Route::from_legs(cities, legs)))
    }

    /// A short, drivable local route from the company terminal to a facility.
    pub fn facility_approach_route(
        &self,
        city: &str,
        location_name: &str,
    ) -> Result<Route, DataError> {
        let city = self.resolve_city_key(city);
        let location = self.facility_location(&city, location_name)?;
        if let Some(source_approach) = self.facility_approaches()?.get(&location.id) {
            if source_approach.turn_level && !source_approach.segments.is_empty() {
                let legs: Vec<Leg> = source_approach
                    .segments
                    .iter()
                    .map(|segment| {
                        Leg::local(
                            &city,
                            segment.miles,
                            &spoken_road_text(&segment.road),
                            &spoken_road_text(&segment.cue),
                            segment.speed_mph,
                        )
                    })
                    .collect();
                let cities = vec![city.clone(); legs.len() + 1];
                return Ok(Route::from_legs(cities, legs));
            }
        }
        let endpoint = self.facility_endpoints()?.get(&location.id);
        let approach = self.local_approach(&format!("facility:{}", location.id))?;
        let (mut miles, road) = match (endpoint, approach) {
            (Some(endpoint), _) if endpoint.source_backed && !endpoint.fallback => (
                endpoint.approach_miles,
                approach
                    .map(|a| a.road.clone())
                    .unwrap_or_else(|| endpoint.approach_road.clone()),
            ),
            (_, Some(approach)) => (approach.approach_miles, approach.road.clone()),
            _ => {
                let base_miles =
                    lookup(FACILITY_APPROACH_MILES, &location.facility_type).unwrap_or(4.0);
                let seed = crc32(
                    format!("{city}:{}:{}", location.name, location.facility_type).as_bytes(),
                );
                let offset = f64::from(seed % 7) * 0.25;
                (
                    round_py_n(base_miles + offset, 1),
                    lookup(FACILITY_APPROACH_ROADS, &location.facility_type)
                        .unwrap_or("facility access road")
                        .to_string(),
                )
            }
        };
        // Sanity clamp for the synthetic single-leg approach: 776 baked
        // records carry up to the bake tool's 35-mile cap because the
        // facility's geocoded pin landed tens of miles from its city --
        // Josh drew a 35-mile straight deadhead in Kenosha (2026-07-24).
        // Until the placement audit re-geocodes them (roadmap), no local
        // deadhead crawls half a county; real multi-leg street chains
        // above are never clamped.
        miles = miles.min(SYNTHETIC_APPROACH_CAP_MI);
        let leg = Leg::new(&city, &city, miles, &road, "flat", Vec::new());
        Ok(Route::from_legs(vec![city.clone(), city], vec![leg]))
    }
}

fn py_key(key: &str) -> String {
    super::world_parsing::py_repr_str(key)
}

#[cfg(test)]
mod tests {
    //! The pure part of `tests/test_street_turns.py`; the world-backed check
    //! lives in `crates/ff-core/tests/data_street_turns.rs`.
    use super::*;

    #[test]
    fn test_spoken_road_text_trims_osm_ref_lists() {
        assert_eq!(
            spoken_road_text("Turn left onto North Michigan Street (SR 933;BUS US 31)."),
            "Turn left onto North Michigan Street (SR 933)."
        );
        assert_eq!(
            spoken_road_text("Slater Street SW (US 19; US 82; GA 3; GA 133; GA 520)"),
            "Slater Street SW (US 19)"
        );
        // Single refs and plain names pass through untouched.
        assert_eq!(
            spoken_road_text("Main Street (SR 26)"),
            "Main Street (SR 26)"
        );
        assert_eq!(
            spoken_road_text("Richard G. Hatcher Boulevard"),
            "Richard G. Hatcher Boulevard"
        );
        // Prose parentheticals with semicolons only lose text after the semicolon,
        // never the sentence around them.
        assert_eq!(
            spoken_road_text("no parens; still fine"),
            "no parens; still fine"
        );
    }

    #[test]
    fn reversed_chain_flips_each_junction() {
        let city = "south_bend_in_us";
        let legs = vec![
            std::sync::Arc::new(Leg::local(
                city,
                0.15,
                "East Navarre Street",
                "Start on East Navarre Street.",
                25.0,
            )),
            std::sync::Arc::new(Leg::local(
                city,
                0.2,
                "North Michigan Street",
                "Turn left onto North Michigan Street.",
                25.0,
            )),
            std::sync::Arc::new(Leg::local(
                city,
                0.5,
                "South Michigan Street",
                "Continue onto South Michigan Street.",
                30.0,
            )),
        ];
        let out = reversed_local_legs(city, &legs);
        let cues: Vec<&str> = out.iter().map(|l| l.local_cue.as_str()).collect();
        assert_eq!(
            cues,
            vec![
                "Start on South Michigan Street.",
                "Continue onto North Michigan Street.",
                "Turn right onto East Navarre Street.",
            ]
        );
        assert_eq!(out[2].local_speed_mph, 25.0);
    }
}
